<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="LongISLND">
    <meta name="author" content="Marghoob Mohiyuddin">

    <title>LongISLND</title>

    <!-- Bootstrap core CSS -->
    <link href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">
    <style>
    .container {
  margin-right: auto;
  margin-left: auto;
  max-width: 760px;
}
    </style>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-58409498-6', 'auto');
  ga('send', 'pageview');

/**
* Function that tracks a click on an outbound link in Google Analytics.
* This function takes a valid URL string as an argument, and uses that URL string
* as the event label.
*/
var trackOutboundLink = function(url) {
   ga('send', 'event', 'outbound', 'click', url, {'hitCallback':
     function () {
     document.location = url;
     }
   });
}

function _gaLt(event){
    var el = event.srcElement || event.target;
    /* Loop up the DOM tree through parent elements if clicked element is not a link (eg: an image inside a link) */
    while(el && (typeof el.tagName == 'undefined' || el.tagName.toLowerCase() != 'a' || !el.href)){
        el = el.parentNode;
    }
    if(el && el.href){
        if(el.href.indexOf(location.host) == -1){ /* external link */
            /* HitCallback function to either open link in either same or new window */
            var hitBack = function(link, target){
                target ? window.open(link, target) : window.location.href = link;
            };
            /* link */
            var link = el.href;
            /* Is target set and not _(self|parent|top)? */
            var target = (el.target && !el.target.match(/^_(self|parent|top)$/i)) ? el.target : false;
            /* send event with callback */
            ga(
                "send", "event", "Outgoing Links", link,
                document.location.pathname + document.location.search,
                {"hitCallback": hitBack(link, target)}
            );
            /* Prevent standard click */
            event.preventDefault ? event.preventDefault() : event.returnValue = !1;
        }
    }
}
/* Attach the event to all clicks in the document after page has loaded */
var w = window;
w.addEventListener ? w.addEventListener("load",function(){document.body.addEventListener("click",_gaLt,!1)},!1)
 : w.attachEvent && w.attachEvent("onload",function(){document.body.attachEvent("onclick",_gaLt)});

</script>
</head>

<body>
    <div class="container" style="padding-bottom: 400px;">
        <div class="page-header">
                <h1>LongISLND</h1>
                <h2>In silico Sequencing of Lengthy and Noisy Datatypes</h2>
        </div>
    
        <div class="jumbotron">
            <p>
                <i>For more information contact us at <a href="mailto:bina.rd@roche.com">bina.rd@roche.com</a></i>
            </p>
            
        </div>
    <h2>Publication [Open access]</h2>

    <div class="panel panel-default" style="font-family:monospace;">
        <div class="panel-body">
            <i>If you use LongISLND in your work, please cite the following:</i><br>
            Bayo Lau, Marghoob Mohiyuddin, John C. Mu, Li Tai Fang, Narges Bani Asadi, Carolina Dallett, and Hugo Y.K. Lam<br>
            <b>LongISLND: In silico Sequencing of Lengthy and Noisy Datatypes</b><br>
            Bioinformatics first published online September 25, 2016
            <a href="http://dx.doi.org/10.1093/bioinformatics/btw602" onclick="trackOutboundLink('http://dx.doi.org/10.1093/bioinformatics/btw602'); return false;">doi:10.1093/bioinformatics/btw602</a>
        </div>
    </div>
<h2>Introduction</h2>
LongISLND is a read simulator which profiles the characteristics of third generation, single-molecule sequencing technologies and simulates accordingly. The general software architecture is easily extendable, as demonstrated by the emulation of Pacific Biosciences (PacBio) multi-pass sequencing with P5 and P6 chemistries, producing data in FASTQ, H5, and the latest PacBio BAM format. Please read on to see application examples to PacBio and oxford nanopre (ONT) data.
<h2>Download LongISLND</h2>
<p>Github repository: <a href="https://github.com/bioinform/longislnd" onclick="trackOutboundLink('https://github.com/bioinform/longislnd'); return false;">https://github.com/bioinform/longislnd</a></p>
<h2>System Requirements</h2>
<p>
   The following must be installed:
   <ul>
      <li>Java 1.7</li>
      <li>Maven (to build from source)</li>
      <li>Python 2.7 (for convinient scripts and usage examples)</li>
      <li>samtools(for convinient scripts and usage examples)</li>
   </ul>
</p>

<h2>Installing LongISLND</h2>
<h4>From Source</h4>
<ul>
  <li>change current working directory into the source code directory</li>
  <li>execute <code>linux_build.sh</code>, which assumes <code>wget</code>, Java 1.7 (<code>java</code>) and Maven (<code>mvn</code>) are available.</li>
  <li>accept HDF Java Products' license and accept default installation location to proceed.</li>
</ul>
<h4>From Binary Download</h4>
<ul>
  <li>download a binary package from <a href="https://github.com/bioinform/longislnd/releases" onclick="trackOutboundLink('https://github.com/bioinform/longislnd/releases'); return false;">https://github.com/bioinform/longislnd/releases</a>
  <li>execute <code>tar xzf ${TAR_GZ}</code> to unpack the tar.gz package
  <li>change current working directory into the binary directory</li>
  <li>execute <code>linux_build.sh</code>, which assumes <code>wget</code> and <code>curl</code> are available.</li>
  <li>accept HDF Java Products' license and accept default installation location to proceed.</li>
</ul>

<h2>Running LongISLND</h2>
<p>Usage of the Java JAR can be carried out/demonstrated by the following convinient Python scripts.</p>
<code>simulate.py</code>
<ul>
  <li>execute with <code>-h</code> for a list of options</li>
  <li>simulate sequences given a model and a genome</li>
</ul>
<code>sample.py</code>
<ul>
  <li>execute with <code>-h</code> for a list of options</li>
  <li>please refer to the Example below for input preparation.</li>
  <li>build model used by <code>simulate.py.</code></li>
  <li>for PacBio's read, the script will look for files whose names are:</li>
  <ul>
    <li>*.fofn, each storing a list of bax.h5 read files</li>
    <li>*.fofn.cmp.h5, each storing the alignment of the corresponding *.fofn</li>
  </ul>
</ul>

<h2>Troubleshooting</h2>
<ul>
  <li>Consider setting maximum memory for java to match that of your computer. For example, <code>java -Xmx20g</code> would allocate 20gb.</li>
  <li>The memory setting can be specified for <code>sample.py</code> and <code>simulate.py</code> with <code> --jvm_opt " -Xmx20g " </code></li>
</ul>

<h2>Usage Examples</h2>
<p>These are demonstrations of LongISLND, from downloading data, to alignment, to learning, and to simulation.</p>
<font color="#FF0000">These are tested only on Linux platform due to various depedencies.</font>
<h3>Installing Aligners</h3>
<ol>
  <h4><li>PacBio</li></h4>
  <p>This sets up PacBio's SMRTAnalysis2.3, if it's not already available to you. All <code>download_and_align.sh</code> scripts in PacBio examples assume an installation location of <code>sampling_example/smrtanalysis</code>, please change those scripts if you want to use your own version of SMARTAnalysis.</p>
  <ol>
    <li>Change working directory to <code>sampling_example</code>.</li>
    <li>execute <code>setup_smrt23.sh</code></li>
    <ul>
      <li>download and build PacBio's SMRTAnalysis2.3</li>
      <li>(to avoid complications) select NONE in answering the question "What job management system will you be using?"
    </ul>
  </ol>
  <h4><li>ONT</li></h4>
  <p>This sets up GraphMap (Sovic, I. et al (2016). Fast and sensitive mapping of nanopore sequencing reads with graphmap. Nat Commun, 7.) and Samtools (H. Li, et al (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics. 25(16)), if it's not already available to you. <code>sampling_example/ont_ecoli/download_and_align.sh</code> assumes installation locations of <code>sampling_example/graphmap</code> and <code>sampling_example/samtools-1.2</code>, please change the script if you want to use your own version of aligners.</p>
  <ol>
    <li>Change working directory to <code>sampling_example</code>.</li>
    <li>execute <code>setup_ont.sh</code></li>
    <ul>
      <li>download and build GraphMap</li>
      <li>download and build Samtools</li>
    </ul>
  </ol>
</ol>

<h3>PacBio P6 E. coli</h3>
<ol>
<li>Change working directory to <code>sampling_example/ecoli</code>.</li>
<li>execute <code>download_and_align.sh</code></li>
<ul>
  <li>download ecoli data and assembly</li>
  <li>align to corresponding assembly</li>
  <li>generate *.fofn and *.fofn.cmp5 used by <code>sample.py</code></li>
</ul>
  <li>execute <code>learn_and_simulate.sh</code></li>
  <ul>
    <li>invoke <code>sample.py</code> to learn model from aligned data</li>
    <li>invoke <code>simulate.py</code> to simulate reads and output in PacBio's BAM format, stored in the default <code>out</code> directory.</li>
  </ul>
</ol>

<h3>ONT R7.3 E. coli</h3>
<ol>
<li>Change working directory to <code>sampling_example/ont_ecoli</code>.</li>
<li>execute <code>download_and_align.sh</code></li>
<ul>
  <li>download ecoli data (Loman, N.J. et al (2015). A complete bacterial genome assembled de novo using only nanopore sequencing data. Nat Meth, 12(8), 733â€“735) </li>
  <li>extract sequences</li>
  <li>align to reference</li>
</ul>
  <li>execute <code>learn_and_simulate.sh</code></li>
  <ul>
    <li>invoke <code>sample.py</code> to learn model from aligned data</li>
    <li>invoke <code>simulate.py</code> to simulate reads and output in the FASTQ format, stored in the default <code>out</code> directory.</li>
  </ul>
</ol>

<h3>PacBio P6 CHM1</h3>
<p>This is storage/io/compute-intensive due to the scale of the human-scale sequencing data.</p>
<ol>
<li>Change working directory to <code>sampling_example/p6_chm1</code>.</li>
<li>execute <code>download_and_align.sh</code></li>
<ul>
  <li>download a subset of P6 CHM1 data</li>
  <li>download the MHAP assembly</li>
  <li>align to corresponding assembly</li>
  <li>generate *.fofn and *.fofn.cmp5 used by <code>sample.py</code></li>
</ul>
  <li>execute <code>learn.sh</code></li>
  <ul>
    <li>learn model from aligned data</li>
    <li>NUM_THREADS can be changed to parallelized learning. At least 3GB per thread is recommended.</li>
  </ul>
  <li>please use <code>simulate.py</code> as in  <code>sampling_example/ecoli/learn_and_simulate.sh</code> to simulate</li>
</ol>


</body>
